### PR TITLE
#161648143 Build Twitter Social Login

### PR DESCRIPTION
**Pull request by mrSerious**
#### What does this PR do?

Add the twitter social login feature

#### Description of Task to be completed?

- add tests
- add twitter routes
- add twitter strategy
- add twitter callback
- set fullName column to accept nulls

#### How should this be manually tested?

Run `npm start` to start the server
Navigate to  [http://localhost:3000/api/v1/auth/twitter](http://localhost:3000/api/v1/auth/twitter) on your browser

#### Any background context you want to provide?

N/A

#### What are the relevant pivotal tracker stories?

[#161648143](https://www.pivotaltracker.com/story/show/161648143)

#### Screenshots (if appropriate)

N/A

#### Questions:

N/A

### CODE REVIEW

File: server/migrations/20181029191911-create-user.js [hunk 1]
```diff
@@ -7,7 +7,7 @@ export default {
       type: Sequelize.INTEGER
     },
     fullName: {
-      allowNull: false,
+      allowNull: true,
```
Comment (mbilesanmi) [inline server/migrations/20181029191911-create-user.js]: Why are we allowing this to be null?

File: server/config/twitterPassportRoutes.js [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+passport.deserializeUser((id, done) => {
+  User.findByPk(id).then((user) => {
+    done(null, user);
+  });
+});
+
+const twitterOptions = {
+  consumerKey: process.env.TWITTER_APP_ID,
+  consumerSecret: process.env.TWITTER_APP_SECRET,
+  callbackURL: 'http://localhost:3000/api/v1/auth/twitter/callback',
```
Comment (mbilesanmi) [inline server/config/twitterPassportRoutes.js]: Do we wanna hard-code this here? What if we are on Production?

File: server/migrations/20181029191911-create-user.js (hunk 1 above)
Comment (mrSerious) [reply server/migrations/20181029191911-create-user.js]: Twitter does not return the first name and last name for users, just the display name and username. I had to make this field allow nulls to allow for creating new Twitter users.

File: server/config/twitterPassportRoutes.js (hunk 2 above)
Comment (mrSerious) [reply server/config/twitterPassportRoutes.js]: No. Can we set the port as an environment variable?

Comment (mbilesanmi) [reply server/config/twitterPassportRoutes.js]: Can't the callbackURL be set as an env and then you call the needed one based on your environment?

Comment (mrSerious) [reply server/config/twitterPassportRoutes.js]: I see what you mean. Should I make this change?

**Comment by KvNGCzA** [comment]
@mbilesanmi LGTM

**Comment by mrSerious** [comment]
@sulenchy we agreed to leave this as is yesterday.

**Pull request merged**